File: 129. Sum Root to Leaf Numbers.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    private List<List<Integer>> res = new ArrayList<>();
    //private List<Integer> temp = new ArrayList<>();
    private int sum = 0;
    
    public int sumNumbers(TreeNode root) {
        dfs(root, new ArrayList<>());
        return sum;
    }
    
    public int lst2int(List<Integer> path) {
        // [1,2,9] -> 129
        StringBuilder sb = new StringBuilder();
        for(int i = 0; i < path.size(); i++) {
            int cur = path.get(i);
            sb.append(Integer.toString(cur));
        }
        String s = sb.toString();
        return Integer.parseInt(s);
    }
    
    // dfs ,backtracking
    public void dfs(TreeNode root, List<Integer> temp) {
        if (root == null) return;
        if (root.left == null && root.right == null) {
            temp.add(root.val);
            List<Integer> tempCp = new ArrayList<>(temp);
            int val = lst2int(tempCp);
            //res.add(val);
            sum += val;
            temp.remove(temp.size() - 1);
            return;
        }
        temp.add(root.val);
        dfs(root.left, temp);
        dfs(root.right, temp);
        temp.remove(temp.size() - 1);
    }
}

// Time: dfs + for each element in each path

File: amazon/python/129. Sum Root to Leaf Numbers.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def sumNumbers(self, root: TreeNode) -> int:
        self.res = 0
        self.dfs(root,0)
        return self.res
    
    def dfs(self, root, value):
        if root:
            if not root.left and not root.right:
                self.res += value * 10 + root.val 
            self.dfs(root.left, value * 10 + root.val)
            self.dfs(root.right, value * 10 + root.val)
            
    # Time: O(n)
    # Space:O(n)
            
        
#         root_to_leaf = 0
#         if root is None:
#             return 0
        
#         stack = [(root,0)]
#         while stack:
#             root, curr = stack.pop()
#             curr = curr * 10 + root.val
#             if not root.left and not root.right:
#                 root_to_leaf += curr
#             if root.left:
#                 stack.append((root.left, curr))
#             if root.right:
#                 stack.append((root.right, curr))
                
#         return root_to_leaf
    
    # Time: O(n)
    # Space: O(n)
    
    
    
    
    
